pageserver/src/tenant/timeline/layer_manager.rs

```diff
@@ -254,40 +269,34 @@ impl LayerManager {
     /// Remote storage is not affected by this operation.
     fn delete_historic_layer(
         // we cannot remove layers otherwise, since gc and compaction will race
-        _layer_removal_cs: Arc<tokio::sync::OwnedMutexGuard<()>>,
-        layer: Arc<dyn PersistentLayer>,
+        _layer_removal_cs: &Arc<tokio::sync::OwnedMutexGuard<()>>,
+        layer: Layer,
         updates: &mut BatchedUpdates<'_>,
-        metrics: &TimelineMetrics,
-        mapping: &mut LayerFileManager,
+        mapping: &mut LayerFileManager<Layer>,
     ) -> anyhow::Result<()> {
         let desc = layer.layer_desc();
-        if !layer.is_remote_layer() {
-            layer.delete_resident_layer_file()?;
-            metrics.resident_physical_size_gauge.sub(desc.file_size);
-        }
 
         // TODO Removing from the bottom of the layer map is expensive.
         //      Maybe instead discard all layer map historic versions that
         //      won't be needed for page reconstruction for this timeline,
         //      and mark what we can't delete yet as deleted from the layer
         //      map index without actually rebuilding the index.
         updates.remove_historic(desc);
-        mapping.remove(layer);
+        mapping.remove(&layer);
+        layer.garbage_collect();
```

Comment:
When glossing over the code, a method named "garbage collect" sounds to my ears like a consuming API for a RAII resource.

I think a more intuitive name would be `try_delete_layer_file_on_drop`.

Reply:
I don't follow. Are you worried the function might be taking a `self`, dropping all of the clones and garbage collecting right away?

Reply:
Renamed it to `garbage_collect_on_drop` in 7c2f687bd. I don't think there's any trying or something like with the action to mark it garbage_collected